wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1388,3 +1396,192 @@ let construct_pointer_constraint (l_orig : Constr.z3_expr list) (env1 : Env.t)
   in
   let expr = List.fold l ~init:(Bool.mk_true ctx) ~f:gen_constr in
   Constr.mk_goal "pointer_precond" expr |> Constr.mk_constr
+
+(* Gets the target destination of a jump. We do not handle indirect jumps. *)
+let target (jmp : Jmp.t) : Tid.t option =
+  match Jmp.kind jmp with
+  | Goto (Direct tid) -> Some tid
+  | Int (_, tid) -> Some tid
+  | Call t -> begin
+      match Call.return t with
+      | Some (Direct tid) -> Some tid
+      | Some (Indirect _) | None -> None
+    end
+  | Goto (Indirect _) | Ret _ -> None
+
+(* Checks if [tid] corresponds to the tid of a block in a loop. *)
+let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
+  Seq.exists (Group.enum loop) ~f:(fun n ->
+      Tid.equal (n |> Graphs.Ir.Node.label |> Term.tid) tid)
+
+(* Assumes the loop header branches into two blocks. *)
+let get_branch (blk : Blk.t) : (Jmp.t * Jmp.t) option =
+  match Seq.to_list (Term.enum jmp_t blk) with
+  | [jmp1; jmp2] -> Some (jmp1, jmp2)
+  | _ -> None
+
+(* Gets the BAP expression that represents the condition for going into the
+   loop. *)
+let get_branch_condition (loop : Graphs.Ir.Node.t group) (jmp1 : Jmp.t)
+    (jmp2 : Jmp.t) : Exp.t option =
+  match target jmp1, target jmp2 with
+  | Some target1, Some target2 ->
+    if in_loop loop target1 then
+      (* jmp1 goes to loop, jmp2 exits. while(E) is jmp1's condition. *)
+      Some (Jmp.cond jmp1)
+    else if in_loop loop target2 then
+      (* jmp1 exits, jmp2 goes to loop. while(E) is the negation of jmp1's
+         condition. *)
+      Some Bil.(unop not (Jmp.cond jmp1))
+    else
+      None
+  | _, _ -> None
+
+(* Gets a constraint representing the condition for a loop. *)
+let get_cond (loop : Graphs.Ir.Node.t group) (blk : Blk.t) (env : Env.t)
+  : Constr.t option =
+  let ctx = Env.get_context env in
+  let* jmp1, jmp2 = get_branch blk in
+  let* branch_cond = get_branch_condition loop jmp1 jmp2 in
+  let cond_val, _, env = exp_to_z3 branch_cond env in
+  let cond_size = BV.get_size (Expr.get_sort cond_val) in
+  let loop_guard =
+    Bool.mk_not ctx (Bool.mk_eq ctx cond_val (z3_expr_zero ctx cond_size))
+    |> Constr.mk_goal "Loop guard"
+    |> Constr.mk_constr
+  in
+  (* Compute to get the constraint at the top of the block. *)
+  let compute elts =
+    Seq.fold elts ~init:(loop_guard, env) ~f:(fun (pre, env) elt ->
+        match elt with
+        | `Def _ | `Phi _ -> visit_elt env pre elt
+        | `Jmp jmp ->
+          if (Jmp.equal jmp jmp1) || (Jmp.equal jmp jmp2) then
+            pre, env
+          else
+            visit_elt env pre elt)
+  in
+  let pre, _ = blk |> Blk.elts ~rev:true |> compute in
+  Some pre
+
+(* Gets the body of a loop without the header node containing the loop
+   condition. *)
+let get_body (graph : Graphs.Ir.t) (loop : Graphs.Ir.Node.t group) (blk : Blk.t)
```

Comment:
We should probably include the header node in the body (without the backedge into the header).

Reply:
Indeed.